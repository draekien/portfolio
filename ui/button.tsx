import { cn } from '@/utils/cn';
import { cva, type VariantProps } from 'class-variance-authority';
import { forwardRef } from 'react';

export const buttonVariants = cva(
  'inline-flex items-center justify-center gap-2 text-sm font-medium disabled:opacity-50 disabled:cursor-not-allowed [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0',
  {
    variants: {
      variant: {
        default: 'bg-primary text-primary-foreground hover:bg-primary/80',
        destructive: 'bg-destructive text-destructive-foreground hover:bg-destructive/80',
        secondary: 'bg-secondary text-secondary-foreground hover:bg-secondary/80',
        ghost: 'hover:bg-primary/20 hover:text-primary-foreground',
        link: 'rounded-none rounded-tl-md rounded-tr-md border-b border-b-primary hover:border-b-primary/80 hover:bg-primary/10',
      },
      size: {
        default: 'h-10 px-4 py-2',
        sm: 'h-9 px-3',
        lg: 'h-11 px-8',
        icon: 'h-10 w-10',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
);

export type ButtonProps = React.ComponentPropsWithoutRef<'button'> &
  VariantProps<typeof buttonVariants>;

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ children, className, size, variant, ...props }, ref) => {
    return (
      <button
        ref={ref}
        className={cn(buttonVariants({ variant, size, className }))}
        {...props}>
        {children}
      </button>
    );
  }
);

Button.displayName = 'Button';

export default Button;
