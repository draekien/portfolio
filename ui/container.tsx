import { cn } from '@/utils/cn';
import { forwardRef } from 'react';

export type ContainerProps = React.ComponentPropsWithoutRef<'div'>;

const Container = forwardRef<HTMLDivElement, ContainerProps>(
  ({ children, className, ...rest }, ref) => (
    <div ref={ref} {...rest} className={cn('container py-4 md:py-8', className)}>
      {children}
    </div>
  )
);

export default Container;
