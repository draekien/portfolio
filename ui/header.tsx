import Image from 'next/image';

export default function Header() {
  return (
    <header className="container sticky top-0 flex items-start place-self-center py-4 md:py-8">
      <div className="flex origin-top-left flex-col text-lg font-medium uppercase md:text-xl">
        <span className="leading-none">William</span>
        <span className="leading-none">Pei</span>
      </div>
      <Image
        src="logo.svg"
        alt="William Pei's logo"
        width={92}
        height={60}
        className="ml-auto aspect-square w-6 origin-top-right md:w-11"
      />
    </header>
  );
}
