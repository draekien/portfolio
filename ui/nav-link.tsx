'use client';

import { cn } from '@/utils/cn';
import * as NavigationMenu from '@radix-ui/react-navigation-menu';
import { usePathname } from 'next/navigation';
import { forwardRef } from 'react';
import Link from './link';

export type NavLinkProps = React.ComponentPropsWithoutRef<typeof Link>;

const NavLink = forwardRef<React.ElementRef<typeof NavigationMenu.Link>, NavLinkProps>(
  ({ className, children, href, size, variant, ...rest }, ref) => {
    const pathname = usePathname();
    const active = pathname === href;

    return (
      <NavigationMenu.Link ref={ref} asChild active={active}>
        <Link
          href={href}
          className={cn(
            'data-[active]:underline data-[active]:decoration-secondary data-[active]:underline-offset-8',
            className
          )}
          size={size}
          variant={variant}
          disabled={active}
          {...rest}>
          {children}
        </Link>
      </NavigationMenu.Link>
    );
  }
);

NavLink.displayName = NavigationMenu.Link.displayName;

export default NavLink;
