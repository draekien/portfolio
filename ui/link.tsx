import { cn } from '@/utils/cn';
import NextLink from 'next/link';
import { forwardRef } from 'react';
import { buttonVariants, type ButtonProps } from './button';

export type LinkProps = React.ComponentPropsWithoutRef<typeof NextLink> &
  Pick<ButtonProps, 'variant' | 'size' | 'disabled'>;

const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  ({ children, variant, size, className, disabled, ...props }, ref) => {
    return (
      <NextLink
        ref={ref}
        className={cn(
          buttonVariants({ variant, size, className }),
          'aria-disabled:pointer-events-none'
        )}
        aria-disabled={disabled}
        tabIndex={disabled ? -1 : undefined}
        {...props}>
        {children}
      </NextLink>
    );
  }
);

Link.displayName = 'Link';

export default Link;
