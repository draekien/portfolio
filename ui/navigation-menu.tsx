import { cn } from '@/utils/cn';
import { ChevronDownIcon } from '@radix-ui/react-icons';
import * as NavigationMenuPrimitive from '@radix-ui/react-navigation-menu';
import { forwardRef } from 'react';
import { buttonVariants } from './button';
import NavLink, { type NavLinkProps } from './nav-link';

const NavigationMenu = forwardRef<
  React.ElementRef<typeof NavigationMenuPrimitive.Root>,
  React.ComponentPropsWithoutRef<typeof NavigationMenuPrimitive.Root>
>(({ children, className, ...props }, ref) => {
  return (
    <NavigationMenuPrimitive.Root
      {...props}
      ref={ref}
      className={cn('relative flex justify-center', className)}>
      {children}
    </NavigationMenuPrimitive.Root>
  );
});

NavigationMenu.displayName = NavigationMenuPrimitive.Root.displayName;

const NavigationMenuList = forwardRef<
  React.ElementRef<typeof NavigationMenuPrimitive.List>,
  React.ComponentPropsWithoutRef<typeof NavigationMenuPrimitive.List>
>(({ children, className, ...props }, ref) => {
  return (
    <NavigationMenuPrimitive.List
      {...props}
      ref={ref}
      className={cn('group flex list-none items-center justify-center gap-2', className)}>
      {children}
    </NavigationMenuPrimitive.List>
  );
});

NavigationMenuList.displayName = NavigationMenuPrimitive.List.displayName;

const NavigationMenuTrigger = forwardRef<
  React.ElementRef<typeof NavigationMenuPrimitive.Trigger>,
  React.ComponentPropsWithoutRef<typeof NavigationMenuPrimitive.Trigger>
>(({ children, className, ...props }, ref) => {
  return (
    <NavigationMenuPrimitive.Trigger
      {...props}
      ref={ref}
      className={cn(
        buttonVariants({ variant: 'ghost' }),
        'group',
        'inline-flex items-center justify-center',
        'data-[active]:bg-primary/50 data-[state=open]:bg-primary/50',
        className
      )}>
      {children}
      <ChevronDownIcon
        className="relative top-[1px] ml-1 h-3 w-3 transition duration-200 group-data-[state=open]:rotate-180"
        aria-hidden={true}
      />
    </NavigationMenuPrimitive.Trigger>
  );
});

NavigationMenuTrigger.displayName = NavigationMenuPrimitive.Trigger.displayName;

const NavigationMenuContent = forwardRef<
  React.ElementRef<typeof NavigationMenuPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof NavigationMenuPrimitive.Content>
>(({ children, className, ...props }, ref) => {
  return (
    <NavigationMenuPrimitive.Content
      {...props}
      ref={ref}
      className={cn(
        'left-0 top-0 w-full md:absolute md:w-auto',
        'data-[motion^=from-]:animate-in data-[motion^=to-]:animate-out data-[motion^=from-]:fade-in data-[motion^=to-]:fade-out data-[motion=from-end]:slide-in-from-right-52 data-[motion=from-start]:slide-in-from-left-52 data-[motion=to-end]:slide-out-to-right-52 data-[motion=to-start]:slide-out-to-left-52',
        className
      )}
    />
  );
});

NavigationMenuContent.displayName = NavigationMenuPrimitive.Content.displayName;

const NavigationMenuViewport = forwardRef<
  React.ElementRef<typeof NavigationMenuPrimitive.Viewport>,
  React.ComponentPropsWithoutRef<typeof NavigationMenuPrimitive.Viewport>
>(({ children, className, ...props }, ref) => {
  return (
    <div className="absolute left-0 top-full flex justify-center">
      <NavigationMenuPrimitive.Viewport
        {...props}
        ref={ref}
        className={cn(
          'origin-top-center relative mt-1.5 h-[var(--radix-navigation-menu-viewport-height)] w-full overflow-hidden rounded-md border',
          'bg-popover text-popover-foreground shadow-lg',
          'data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-90 md:w-[var(--radix-navigation-menu-viewport-width)]',
          className
        )}
      />
    </div>
  );
});

NavigationMenuViewport.displayName = NavigationMenuPrimitive.Viewport.displayName;

const NavigationMenuIndicator = forwardRef<
  React.ElementRef<typeof NavigationMenuPrimitive.Indicator>,
  React.ComponentPropsWithoutRef<typeof NavigationMenuPrimitive.Indicator>
>(({ children, className, ...props }, ref) => {
  return (
    <NavigationMenuPrimitive.Indicator
      {...props}
      ref={ref}
      className={cn(
        'data-[state=visible]:animate-in data-[state=hidden]:animate-out data-[state=hidden]:fade-out data-[state=visible]:fade-in top-full z-[1] flex h-1.5 items-end justify-center overflow-hidden',
        className
      )}>
      <div className="bg-border relative top-[60%] h-2 w-2 rotate-45 rounded-tl-sm shadow-md" />
    </NavigationMenuPrimitive.Indicator>
  );
});

NavigationMenuIndicator.displayName = NavigationMenuPrimitive.Indicator.displayName;

const NavigationMenuItem = forwardRef<
  React.ElementRef<typeof NavigationMenuPrimitive.Item>,
  React.ComponentPropsWithoutRef<typeof NavigationMenuPrimitive.Item>
>(({ children, className, ...props }, ref) => {
  return <NavigationMenuPrimitive.Item {...props} ref={ref} className={cn(className)} />;
});

NavigationMenuItem.displayName = NavigationMenuPrimitive.Item.displayName;

export {
  NavigationMenu,
  NavigationMenuContent,
  NavigationMenuIndicator,
  NavigationMenuItem,
  NavLink as NavigationMenuItemLink,
  NavLinkProps as NavigationMenuItemLinkProps,
  NavigationMenuList,
  NavigationMenuTrigger,
  NavigationMenuViewport,
};
