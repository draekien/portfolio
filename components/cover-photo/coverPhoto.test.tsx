import { render } from '@testing-library/react';
import { Waystone } from '@waystone/ui';
import CoverPhoto from './coverPhoto';

describe('CoverPhoto', () => {
  it('should match snapshot', () => {
    const { container } = render(<CoverPhoto src="https://example.com" alt="example" />, {
      wrapper: Waystone,
    });

    expect(container).toMatchSnapshot();
  });
});
