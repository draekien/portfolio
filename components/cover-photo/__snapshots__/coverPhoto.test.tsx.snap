// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoverPhoto should match snapshot 1`] = `
<div>
  <div
    class="css-hz73ym-CoverPhoto"
  >
    <div
      class="css-1p96nn1-CoverPhoto"
    >
      <img
        alt="example"
        data-nimg="fill"
        decoding="async"
        fetchpriority="high"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Fexample.com&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fexample.com&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fexample.com&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fexample.com&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fexample.com&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fexample.com&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fexample.com&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fexample.com&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fexample.com&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: none; object-position: 50% 0%; color: transparent;"
      />
    </div>
  </div>
</div>
`;
